bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/StateManagerTest.java

```diff
@@ -0,0 +1,258 @@
+        }, NullStatsLogger.INSTANCE);
+
+        stateManager.initState();
+        stateManager.registerBookie(true).get();
+        assertTrue(stateManager.isRunning());
+
+        stateManager.transitionToReadOnlyMode().get();
+        // stateManager2 will shutdown
+        assertFalse(stateManager.isRunning());
+        // different dimension of bookie state: running <--> down, read <--> write, unregistered <--> registered
+        // bookie2 is set to readOnly when shutdown
+        assertTrue(stateManager.isReadOnly());
+    }
+
+    @Test
+    public void testReadOnlyBookieTransitions() throws Exception{
+        // readOnlybk, which use override stateManager impl
+        File tmpDir = createTempDir("stateManger", "test-readonly");
+        final ServerConfiguration readOnlyConf = TestBKConfiguration.newServerConfiguration();
+        readOnlyConf.setJournalDirName(tmpDir.getPath())
+                .setLedgerDirNames(new String[] { tmpDir.getPath() })
+                .setJournalDirName(tmpDir.toString())
+                .setZkServers(zkUtil.getZooKeeperConnectString())
+                .setForceReadOnlyBookie(true);
+        ReadOnlyBookie readOnlyBookie = new ReadOnlyBookie(readOnlyConf, NullStatsLogger.INSTANCE);
+        readOnlyBookie.start();
+        assertTrue(readOnlyBookie.isRunning());
+        assertTrue(readOnlyBookie.isReadOnly());
+
+        // transition has no effect if bookie start with readOnly mode
+        readOnlyBookie.getStateManager().transitionToWritableMode().get();
+        assertTrue(readOnlyBookie.isRunning());
+        assertTrue(readOnlyBookie.isReadOnly());
+        readOnlyBookie.shutdown();
+
+    }
+
+    /**
+     * Verify the bookie reg.
+     */
+    @Test
+    public void testRegistration() throws Exception {
+        BookieStateManager stateManager = new BookieStateManager(conf, rm);
+        rm.initialize(conf, () -> {
+            stateManager.forceToUnregistered();
+            // schedule a re-register operation
+            stateManager.registerBookie(false);
+        }, NullStatsLogger.INSTANCE);
+        // simulate sync shutdown logic in bookie
+        stateManager.setShutdownHandler(new StateManager.ShutdownHandler() {
+            @Override
+            public void shutdown(int code) {
+                try {
+                    if (stateManager.isRunning()) {
+                        stateManager.forceToShuttingDown();
+                        stateManager.forceToReadOnly();
+                    }
+
+                } finally {
+                    stateManager.close();
+                }
+            }
+        });
+        stateManager.initState();
+        // up
+        assertTrue(stateManager.isRunning());
+        // unregistered
+        assertFalse(stateManager.isRegistered());
+
+        stateManager.registerBookie(true).get();
+        // registered
+        assertTrue(stateManager.isRegistered());
+        stateManager.getShutdownHandler().shutdown(ExitCode.OK);
```

Comment:
This looks strange to me.
You are calling directly the callback, as far as I can see from this test you can simply run the code in the callback.
If you expect the callback to be called more than once please add some logic to test it, like an AtomicInteger counter

Reply:
I used apis the stateManager exposes to outside, where it will call the actual action methods. And the  returned Future<Void> of public api methods is just an indicator for progress, where I can't pass callback to it. I thought the test cover the logic of the class. Sorry for that I couldn't understand your concern.

Reply:
I can't understand why you have to set the shutdown handler. The code you are passing is not called by the bookie but by the test code itself.
It seems that is testing the shutdown handler but it is not doing it actually

Reply:
I can't understand why you have to set the shutdown handler. The code you are passing is not called by the bookie but by the test code itself.
It seems that is testing the shutdown handler but it is not doing it actually

Reply:
yeah, you're right. The logic of the shutdownHandler is same with the bookie. After we extract the bookie from the tests, I think we can only test the logic by this way although it seems a bit superfluous.